### ISSUE TITLE
Remove all the queries from the node in favor of a mid layer application

### ISSUE THREAD

**Issue by RiccardoM** [desmos-labs/desmos#60]
## Context
Yesterday, during a talk with @angelorc from Bitsong, we were confronting each other on chain state approaches that would allow our chains to run smoothly when the following question arose: 

> Why are we creating REST queries endpoints on full nodes? 

To be honest, I really don't get it too. 

Currently we've built them so that clients can fetch data without having to look at blocks or transactions directly. In the future, however, this might become a problem if the set of data becomes so large that queries can potentially block a node from operating normally. 

The main problem of the on-node queries approach is that nodes **do not** perform index-based queries. Instead, they iterate over the whole set of data and filter out the objects that do not satisfy a given filter. This is highly inefficient and can lead to a lot of problems if the data to filter is really large and the filtering process takes a lot of time. 

As a comparison: 

* on Twitter users create more than **200 billion tweets per year** ([Source](https://www.internetlivestats.com/twitter-statistics/));

* on Facebook every **60 seconds**: 510,000 comments are posted, 293,000 statuses are updated, and 136,000 photos are uploaded ([Source](https://zephoria.com/top-15-valuable-facebook-statistics/))

Assuming to have even **1/100th of just Twitter traffic**, this would mean that we would have to handle 2 billion posts per year. By the end of the first year, assuming 1 microsecond to iterate over 1 post, filtering through all the posts could then take up to 2000 seconds, which are **33 minutes**!

## Middle layer storage and querier
In order to solve this problem, Angelo came up with a really good idea that in my opinion is the one we should adopt too, helping him realizing it. 

The solution is pretty simple: 

> Delete all the module-specific query endpoints from the nodes and create a middle layer data storage and querier application based on real databases

The particularity in this approach lies on the fact that we will use the blockchain only as the **state** and **not** the database. Instead, the database will be created on top of the state by parsing the transactions and reading the data inside them. 

Here is a schema of a possible system: 

![Untitled Diagram](https://user-images.githubusercontent.com/9119016/71240449-70a8a600-2309-11ea-9a3e-bd5cc008aa35.png)

As you can see, the middle layer is divided into the parser, the database and the APIs parts. 

### Parser
It's job should be to read the genesis of the chain and any consequent transaction that is performed inside it. It should fetch such data and store them into a database that allows for easier queries. 

### Database
The database should just hold data inside an indexes-based system. In this case, the best solution available AFAIK is a Mongo replica set which allows for easy clustering and high fault tolerance. 

### APIs
The APIs should provide easy data access to clients that request them. Two possible solutions can be provided: GraphQL, with real-time endpoints that updates once a new data is inserted inside the chain, and REST APIs**, that allow for easy and broadly usable data queries.

## Decentralization
Although this might not seem the best way to decentralize things, this whole system should **not** be the only one that acts as described. However, it should instead be just a reference implementation of how things should be done for people wanting to base their application on Desmos, and should be **100% open source** so that new projects can easily fork any part of it and create their own middle layer application the best way they want. 

## Conclusion
I personally think this system should be the way to go for our use case, allowing clients to read data easier that they do right now and in a most efficient way. However, I'd like to gather any feedback on such system and what do you think could be changed or improved about it @kwunyeung @bragaz

**Comment by bragaz** [comment]
I think the problem you and @angelorc raised on yesterday's conversation is something that's crucial while building long term systems as desmos, bitsong, but also any other chain-based project.
 As we don't have any sort of background to understand how those systems will work with such a huge amount of data, the only thing we can do is relay on how the SDK performs queries on them and, as you’ve made noticed, it seems to do that not in the best way. So I'm agreeing with the solution you've proposed but, as this is something above DESMOS, I think we can handle it later on since the two things (Desmos and the QES*) are on different levels. I think that how graphQL handles data is amazing and someone who will build a desmos based social network could takes benefits only by using it.
For now i will not remove the REST queries, as they are useful for these first phases where data will not be as many as in main net, for testing. 
When the time come, I would be pleased to study how can we implement such a solution as I think this is a good way to help other devs to build great apps on it.

*QES stands for Query Enhancer System, i’ve take the right to choose a name for it  🤣

**Comment by kwunyeung** [comment]
I would say this is not the right approach on the chain itself. You are considering the whole server-client decentralized application architecture but not the architecture of the chain itself.

The architecture suggested is what we are working on Big Dipper. There are different parts we are running on production. I take Cosmos Hub as an example.

1. `Cosmos Hub`  - the blockchain
2. `Curator` - it reads data from the chain and store on MongoDB
3. `Database` - a MongoDB database that store data from the Curator and as a datasource for the frontend the read from.
4. `Web frontend` - show data from the database, interact with users and send txs directly to the Cosmos Hub

I built Big Dipper in this approach as the original `Cosmos Explorer` (the origin of `Lunie.io`) was too slow and you can't search anything on it, there is no data indexed at all. This architecture was borrowed after a DApp on LOOM network called DelegateCall.com which is now dead. There is a medium article about this DApp. https://medium.com/loom-network/announcing-delegatecall-com-the-first-dappchain-live-on-loom-network-f95912388a4c . If we build the DApps in the traditional way, all data just query directly from the chain, the user experience will be very bad. I'm not sure if you had come across any shopping site on Ethereum. You wait for the data to list out the items, wait Metamask to transact, confirm the transaction, etc. The experience is just painful and you won't want to buy things there.

I had the idea of building a responsive block explorer after reading this. Of course original the idea was to monitor our own development chain only 🤣 

There is an architecture diagram and I put it here as reference.

![](https://miro.medium.com/max/1024/1*gHyY25cIMI_mkoFm6nlChA.png)

DelegateCall.com used MySQL as database server and there was a Redis as caching. 

If you compare what you have suggested, here are the mapping.

| Big Dipper        | DelegateCall           | Your scenario  |
| ------------- |-------------| -----|
| Cosmos Hub      | DelegateCall DApp Blockchain | The Blockchain |
| Curator      | Worker in DelegateCall      |   Parser  |
| MongoDB | MySQL database + Redis     |    MongoDB replica set |
| Web frontend | DelegateCall explorer | Client |

In this architecture, the query API from the chain can't be removed as there must be a way for the middle layer to read data from. Of course we can read directly from the blocks with `Tendermint` RPC but then we need to write decoders for each message in different module in the parser which is not desirable for other developers or simple readers. The RESTful API in the modules now serve the purpose for the Parser to read the data and it can't be totally removed. The issue #11 I suggested was to replace the RESTful API as it's too clumsy to get data by each endpoint. If it's possible to read the data by requesting a query and get the data we want, we don't have to develop so many endpoints. I haven't dug deep in this topic yet but the reference I put on the issue #11 is a library of GraphQL in Go which I think if we can define the schema and link it with the KVStore, the data can be queried out. I have discussed a little bit with Bez before and I have also sent this library to him. I think they haven't thought of this before as he asked me if this is possible. It seems that he had studied a bit and found it's possible as he has recently wanted to propose a proposal of building a Cosmos SDK module which can query the data on the chain with GraphQL.

I have been asked many times if Big Dipper has an API but we are not serving yet as I am hesitate to make Big Dipper to become the single point of trust. Although we have some metrics and aggregated data, it's lovely to let other apps to get these data and show them in their own ways. But they always can't be verified and proved. How do you know if the average delegation ratio I have calculated for yesterday is correct and trust worthy? Meteor integrates GraphQL smoothly and they are the company behind [Apollo](https://www.apollographql.com/). We can just add a package and all data can be queried.

In short, the query API in the modules can't be removed. Instead, I prefer a unify way like GraphQL to query but not building every single function for each endpoint. The caching, data aggregation, etc. should leave for DApp developers to design their architecture.

**Comment by kwunyeung** [comment]
There is a project on Steem which is similar to this idea. 

https://github.com/SteemData/steemdata

**Comment by RiccardoM** [comment]
@kwunyeung @bragaz I think it's about time we discuss further on this issue and the associated #11. 

Particularly, we now need to face the problem that might raise in the near future if a lot of user suddenly start using Mooncake and the volume of posts start raising. 

Currently both the Android and iOS applications fully rely on the same now both for: 

1. querying old data using the `/posts` endpoint; and
2. getting new data subscribing to the websocket. 

What I think we should do is implement a system architecture like the one described in [@kwunyeung's previous comment](https://github.com/desmos-labs/desmos/issues/60#issuecomment-568428196). In particular, here's what I think should be created.

## Schema
![Desmos querier system (1)](https://user-images.githubusercontent.com/9119016/74935525-ac82e900-53e8-11ea-8896-ea19ac3e2a49.png)

## Parts descriptions
### Parser
The parser fetches the data from the blockchain saving it inside a MongoDB replica set. Here is how it should act:

1. The first time it connects, it downloads all the data using the REST APIs exposed from fullnodes (like the `/posts` and `/sessions` ones). 

2. Once it has downloaded all the current data, it should start observing for new data by subscribing the websocket provided from the fullnodes. Upon each event, it should execute the proper actions in order to update the stored data. 

This is pretty much the same thing that currently happens inside the Mooncake mobile application, so it can be replicated pretty easily and should not overload the fullnode by a millions of requests, nor it should overload the parser itself with hard operations (i.e block parsing) as the data retrieved from the REST APIs will be already properly formatted. 

### Web server
The web server exposes to the clients the REST and GrapQL endpoints so that they can fetch the data in the best form they which so. 

### MongoDB replica set  
This is where the data is stored. 

## Notes
I think that having a replica set and a web server instead of directly implementing GraphQL APIs on the node itself (as described inside #11), will be much more efficient as current Golang databases perform much more slower than MongoDB when dealing with object storing, a lot of reads and complex queries. 

Currently MongoDB is ranked n.5 inside the [DB-Engines ranking](https://db-engines.com/en/ranking), and it has shown to be one of the fastest alternatives for NoSQL storing. 

Apart from this, we could use [Apollo](https://www.apollographql.com/docs/) as suggested by @kwunyeung, which allows for easy [GraphQL implementation](https://www.apollographql.com/docs/) and also supports [subscriptions](https://www.apollographql.com/docs/react/data/subscriptions/). 

This should allow us to build such system pretty easily to be honest.

**Comment by kwunyeung** [comment]
For performance wise, this is a lot faster than requesting to the chain directly. Maybe we should run something similar to the SteemData project. 

I'm also planning a GraphQL endpoint using Apollo on Big Dipper so that the public can retrieve the aggregated data on Big Dipper. Yes, the beauty of Apollo is that they support subscription. Data are reactive with Apollo client.

BTW, Angelo just added me to his private repo which is a Desmos parser last night. 😆

**Comment by bragaz** [comment]
I think this is the best architecture also for the long term. Should we schedule it for the next version or the one after that?

**Comment by kwunyeung** [comment]
Seems more feasible for the one version after. And I think @MonikaCat can help as she is currently setting up the GraphQL service for Big Dipper.

**Comment by RiccardoM** [comment]
@kwunyeung I'm actually working together with @angelorc to the parser, and I had the chance to improve the code today. 

I think that once that parser is done, If @MonikaCat will be able to give us some advice the hole project can be done in time for `v0.4.0`.

**Comment by angelorc** [comment]
Hi there, yes it seem that I'm involved with @RiccardoM to the `parser` and `gql` :D

Based on my experience with bitsong there are a lot of way to implement `gql`, I'm really curious to know how @MonikaCat wants to implement the data structure

**Comment by RiccardoM** [comment]
## Update
Now that me and @angelorc have built the [Juno fork](https://github.com/desmos-labs/juno), I think we are ready to implement a middle-layer GraphQL server for Desmos as well. 

What I propose to do is the following: 

1. Use the [Juno](https://github.com/desmos-labs/juno) repository as a library to build a custom Desmos Juno implementation.   
   On each transaction, this custom implementation should properly handle posts, users and other data updates. this can be done using the `worker.RegisterMsgHandler` method.

2. Use [Hasura](https://hasura.io/) to build a GraphQL server.  
   This server should allow clients to easily fetch posts, users data and all the other data that might be useful inside a Desmos mobile client. 

By completing the two points above, I think we should have a production-ready GraphQL server in not much time, and we should be able to start testing it inside Mooncake as well.

**Comment by bragaz** [comment]
Personally I think this is the perfect moment to get our hands dirty on this, doing this now will help us to properly understand GraphQL while we are still in testnet. 
I think this approach will also give us enough time to test this kind of solution to see if it will pay in long term and also to build a sort of "log" where we can register how the whole system reacts to a growing amount of data.
I'm going to try to use Juno to build the custom Desmos implementation from today if you are okay with this, then probably move to GraphQL.

**Comment by RiccardoM** [comment]
@bragaz I'm ok with you building the custom Juno implementation for Desmos, as well as later trying to use Hasura for the GraphQL part.

**Comment by kwunyeung** [comment]
Great! Sounds good to me. Hasura looks promising and we should try. If it works well, it may become a standard way of building dapps on Desmos and around Cosmos chains.

**Comment by RiccardoM** [comment]
The whole system has been implement inside [Djuno](https://github.com/desmos-labs/djuno).
